### PR TITLE
Remote caller

### PR DESCRIPTION
**Pull request by apriendeau**
This fixes the problems in #312 to address #63.

### CODE REVIEW

**Comment by apriendeau** [comment]
//cc @aybabtme

**Comment by aybabtme** [comment]
thanks @apriendeau, I'll have a look. I don't know if you wrote all of this or if it's reusing #312, but it will need a decent amount of work before being merged. If you're reusing #312, I'd recommend maybe taking a fresh stab.

Anyhow I'll proceed with a review.

File: caller_test.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+type Output struct {
+	caller string `json:"caller"`
+	msg    string `json:"msg"`
+}
+
+func TestEntryTEXTFormatterCaller(t *testing.T) {
+	fileName := "caller.txt"
```
Comment (aybabtme) [inline caller_test.go]: don't use a file, use a `bytes.Buffer` instead.

File: entry.go [hunk 2]
```diff
@@ -25,20 +25,29 @@ type Entry struct {
 	Time time.Time
 
 	// Level the log entry was logged at: Debug, Info, Warn, Error, Fatal or Panic
-	Level Level
+	Level
```
Comment (aybabtme) [inline entry.go]: don't embed the Level

File: entry.go [hunk 3]
```diff
@@ -75,12 +84,13 @@ func (entry *Entry) WithFields(fields Fields) *Entry {
...
 		data[k] = v
 	}
-	return &Entry{Logger: entry.Logger, Data: data}
+	return &Entry{Logger: entry.Logger, Data: data, depth: entry.depth}
 }
 
 // This function is not declared with a pointer value because otherwise
 // race conditions will occur when using multiple goroutines
 func (entry Entry) log(level Level, msg string) {
+	//	caller(&msg, entry.depth, entry.Logger.showCaller)
```
Comment (aybabtme) [inline entry.go]: remove commented out code

File: entry.go [hunk 4]
```diff
@@ -121,7 +131,9 @@ func (entry *Entry) Debug(args ...interface{}) {
 }
 
 func (entry *Entry) Print(args ...interface{}) {
-	entry.Info(args...)
+	if entry.Logger.Level >= InfoLevel {
```
Comment (aybabtme) [inline entry.go]: why is this changing? seems out of scope

**Comment by aybabtme** [comment]
There's 2 binaries that are committed in this patch, please change the history to remove them.

File: caller_test.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+			}
+		}
+		assert.Equal(t, msg, call)
+	}
+	err = os.Remove(fileName)
+	assert.NoError(t, err)
+}
+
+func TestEntryJSONFormatterCaller(t *testing.T) {
+	fileName := "caller.json"
```
Comment (aybabtme) [inline caller_test.go]: same thing here, don't create files in tests, instead use a `bytes.Buffer`

File: examples/caller/main.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
+// main
```
Comment (aybabtme) [inline examples/caller/main.go]: this example is unnecesary

File: examples/hook/hook.go [hunk 7]
```diff
@@ -2,7 +2,7 @@ package main
 
 import (
 	"github.com/Sirupsen/logrus"
-	"gopkg.in/gemnasium/logrus-airbrake-hook.v2"
+	airbrake "gopkg.in/gemnasium/logrus-airbrake-hook.v2"
```
Comment (aybabtme) [inline examples/hook/hook.go]: why this change?

File: json_formatter.go [hunk 8]
```diff
@@ -11,7 +11,7 @@ type JSONFormatter struct {
 }
 
 func (f *JSONFormatter) Format(entry *Entry) ([]byte, error) {
-	data := make(Fields, len(entry.Data)+3)
+	data := make(Fields)
```
Comment (aybabtme) [inline json_formatter.go]: why this change?

File: logger.go [hunk 9]
```diff
@@ -21,13 +21,15 @@ type Logger struct {
 	// logs with colors, but to a file it wouldn't. You can easily implement your
 	// own that implements the `Formatter` interface, see the `README` or included
 	// formatters for examples.
-	Formatter Formatter
+	Formatter
```
Comment (aybabtme) [inline logger.go]: don't embed things

File: logger.go [hunk 10]
```diff
@@ -21,13 +21,15 @@ type Logger struct {
...
 	// The logging level the logger should log at. This is typically (and defaults
 	// to) `logrus.Info`, which allows Info(), Warn(), Error() and Fatal() to be
 	// logged. `logrus.Debug` is useful in
-	Level Level
+	Level
```
Comment (aybabtme) [inline logger.go]: same here

File: text_formatter.go [hunk 11]
```diff
@@ -57,8 +57,21 @@ type TextFormatter struct {
 }
 
 func (f *TextFormatter) Format(entry *Entry) ([]byte, error) {
-	var keys []string = make([]string, 0, len(entry.Data))
-	for k := range entry.Data {
+	data := make(Fields)
+	for k, v := range entry.Data {
```
Comment (aybabtme) [inline text_formatter.go]: those changes are unrelated to the purpose of the PR

File: logger.go [hunk 12]
```diff
@@ -44,10 +46,11 @@ type Logger struct {
...
 	return &Logger{
-		Out:       os.Stderr,
-		Formatter: new(TextFormatter),
-		Hooks:     make(LevelHooks),
-		Level:     InfoLevel,
+		Out:        os.Stderr,
+		Formatter:  new(TextFormatter),
+		Hooks:      make(LevelHooks),
+		Level:      InfoLevel,
+		showCaller: true,
```
Comment (aybabtme) [inline logger.go]: this will change the output of existing users, the default should be false

File: formatter.go [hunk 13]
```diff
@@ -30,16 +35,33 @@ type Formatter interface {
 //
 // It's not exported because it's still using Data in an opinionated way. It's to
 // avoid code duplication between the two default formatters.
-func prefixFieldClashes(data Fields) {
-	if t, ok := data["time"]; ok {
-		data["fields.time"] = t
+func prefixFieldClashes(data Fields, showCaller bool, depth int) {
+	if _, ok := data["time"]; ok {
+		data["fields.time"] = data["time"]
```
Comment (aybabtme) [inline formatter.go]: those `_, ok := data["time"]` changes are out of scope

File: caller_test.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+		var output Output
+		err = json.Unmarshal([]byte(ss), &output)
+		assert.NoError(t, err)
+		assert.Equal(t, output.caller, output.msg)
+	}
+	err = os.Remove(fileName)
+	assert.NoError(t, err)
+}
+
+func check(l *Logger) {
```
Comment (aybabtme) [inline caller_test.go]: this check should be rolled into a form of table testing, the intent is lost in the length of the function body, and it's hard to grasp what's being tested

**Comment by apriendeau** [comment]
@aybabtme @kpurdon Removed  all the out of scope comments and fixed all the suggestions.

**Comment by aybabtme** [comment]
@apriendeau hey I haven't forgot about you, just haven't had a chance to look into it yet. I'll come back to it sometime this week.

**Comment by apriendeau** [comment]
@aybabtme no worries. We all lead normal lives too 👍  I know everyone at Bitly is chomping at the bit for this feature. 😆

**Comment by apriendeau** [comment]
@aybabtme Any possible way that I can nudge you a little bit on this?

File: logger.go [hunk 15]
```diff
@@ -44,10 +46,11 @@ type Logger struct {
...
 	return &Logger{
-		Out:       os.Stderr,
-		Formatter: new(TextFormatter),
-		Hooks:     make(LevelHooks),
-		Level:     InfoLevel,
+		Out:        os.Stderr,
+		Formatter:  new(TextFormatter),
+		Hooks:      make(LevelHooks),
+		Level:      InfoLevel,
+		showCaller: false,
```
Comment (bobbytables) [inline logger.go]: Any reason this can't be exported?

Comment (apriendeau) [inline logger.go]: I don't see any good reason why it shouldn't be.

**Comment by xuxiangyang** [comment]
Hi, Is the PR going to merge?

**Comment by apriendeau** [comment]
@Sirupsen @aybabtme Where are we on getting this merged?

**Comment by xuxiangyang** [comment]
@apriendeau @Sirupsen @aybabtme Hi， Is the the PR going to merge? Something went wrong? Any reply please. This feature is  badly in need~

**Comment by apriendeau** [comment]
I am sitting in the same boat. I hope to hear something soon @xujinzheng

**Comment by xujinzheng** [comment]
@apriendeau, I think this is good idea to add caller info into log `Entry`, if this PR merged, I will upgrade my `logrus_mate`

**Comment by apremalal** [comment]
IMHO This is a much needed feature for logrus @Sirupsen @apriendeau @aybabtme , appreciate if you guys could merge the PR.

**Comment by joelnb** [comment]
Great work on this, it's a really useful feature!

Is it correct that the caller is only added automatically when using a `JSONFormatter`? I was trying to get this PR working and found that I only saw the caller information when I changed the formatter. Looking at the code it seems correct that 'caller' is never called when using a `TextFormatter` & `TestEntryTEXTFormatterCaller` uses the (unexported) 'caller' method directly.

File: text_formatter.go [hunk 16]
```diff
@@ -68,8 +68,6 @@ func (f *TextFormatter) Format(entry *Entry) ([]byte, error) {
 
 	b := &bytes.Buffer{}
 
-	prefixFieldClashes(entry.Data)
```
Comment (joelnb) [inline text_formatter.go]: Should this be `prefixFieldClashes(data, entry.Logger.showCaller, entry.depth)` to match the change in `JSONFormatter`?

**Comment by horkhe** [comment]
@apriendeau thank you for this change. On question though, why did no you just added `File` and `LineNo` fields to `logrus.Entry`? In that case any hook implementation could have used it as they see fit.

**Comment by apriendeau** [comment]
@horkhe So that it could be apart of the log line itself. We have talked about it being it fields as well.

File: formatter.go [hunk 17]
```diff
@@ -30,16 +35,31 @@ type Formatter interface {
...
+
+	if showCaller {
+		if _, ok := data["caller"]; ok {
+			data["fields.caller"] = data["caller"]
+		}
+		data["caller"] = caller(depth)
+	}
+}
+
+func caller(depth int) (str string) {
```
Comment (puma252) [inline formatter.go]: Depth  hard code into Entry struct. logrus.Entry.log()  (entry.go line: 92) invoke entry.Logger.Hooks.Fire() and entry.Reader(). The depth of  entry.Logger.Hooks.Fire()  different from entry.Reader(). Function caller() not cover the case. I give a patch below, hope it's useful.

`func prefixFieldClashes(data Fields, showCaller bool, depth int) {
	if t, ok := data["time"]; ok {
		data["fields.time"] = t
	}
	if m, ok := data["msg"]; ok {
		data["fields.msg"] = m
	}
	if l, ok := data["level"]; ok {
		data["fields.level"] = l
	}

	if showCaller {
		if _, ok := data["caller"]; ok {
			data["fields.caller"] = data["caller"]
		}

		data["caller"] = getcaller(depth)
	}
}

func caller(depth int) (str string) {

	_, file, line, ok := runtime.Caller(depth)
	if !ok {
		str = "???: ?"
	} else {
		str = fmt.Sprint(filepath.Base(file), ":", line)
	}
	return
}

func getcaller(depth int) (str string) {

	MaxDepth := 9 //max search depth
	UnKnownFileInfo := "???: ?"

	d := depth - 2 //the min depth of logrus.Entry.log
	funcName := ""
	str = UnKnownFileInfo

	pc, _, _, ok := runtime.Caller(d)

	for {

		if !ok {
			return
		}
		funcName = runtime.FuncForPC(pc).Name()
		funcNames := strings.Split(funcName, "/")

		if funcNames[len(funcNames)-1] == "logrus.Entry.log" {
			str = caller(d + 3)
			return
		}

		d = d + 1
		pc, _, _, ok = runtime.Caller(d)

		if d > MaxDepth {
			fmt.Printf("Failed to get File infomation, execeed the max depth %d\n", MaxDepth)
			break
		}
	}

	return
}`

File: formatter.go [hunk 18]
```diff
@@ -30,16 +35,31 @@ type Formatter interface {
...
 	}
+
+	if showCaller {
+		if _, ok := data["caller"]; ok {
+			data["fields.caller"] = data["caller"]
+		}
+		data["caller"] = caller(depth)
+	}
+}
+
```
Comment (puma252) [inline formatter.go]: Depth  hard code into Entry struct. logrus.Entry.log()  (entry.go line: 92) invoke entry.Logger.Hooks.Fire() and entry.Reader(). The depth of  entry.Logger.Hooks.Fire()  different from entry.Reader(). Function caller() not cover the case. I give a patch below, hope it's useful.
[https://github.com/puma252/patchtopull349/blob/master/formatter.go](url)
